scarlet/source.py

```diff
@@ -275,41 +248,36 @@ def init_multicomponent_source(sky_coord, frame, observation, bg_rms, flux_perce
         morphs[k] /= morphs[k].max()
 
     # optimal SEDs given the morphologies, assuming img only has that source
-    seds = get_best_fit_seds(morphs, frame, observation)
+    boxed_img = bbox.image_to_box(observations[obs_idx].images)
+    seds = get_best_fit_seds(morphs, frame, boxed_img)
 
     for k in range(K):
-        if np.any(seds[k] <= 0):
+        if np.all(seds[k] <= 0):
             # If the flux in all channels is  <=0,
             # the new sed will be filled with NaN values,
             # which will cause the code to crash later
             msg = "Zero or negative SED {} for component {} at y={}, x={}".format(seds[k], k, *sky_coord)
-            if np.all(sed <= 0):
-                logger.warning(msg)
-            else:
-                logger.info(msg)
+            logger.warning(msg)
 
-    return seds, morphs
+    return seds, morphs, bbox
 
 
-class RandomSource(Component):
-    """Sources with uniform random morphology.
+class RandomSource(FactorizedComponent):
+    """Sources with uniform random morphology and sed.
 
     For cases with no well-defined spatial shape, this source initializes
     a uniform random field and (optionally) matches the SED to match a given
     observation.
     """
-
-    def __init__(self, frame, observation=None, **component_kwargs):
-        """Source intialized with a single pixel
+    def __init__(self, frame, observation=None):
```

Comment:
+ 1

Reply:
We don't allow passing a constraint into the source classes elsewhere, and there's a reason for that. I see the source classes as the way of specifying models in a reproducible fashion. Think repository. That means they need to include everything. It would be problematic to pickle the sources otherwise.

In essence, if you want a `RandomSource` with different constraints, simply create `DifferentRandomSource`.

Reply:
I agree with you, and yet `ExtendedSource` takes `symmetric` and `monotonic` arguments to specify the constraints. So by that argument we should use the (new) defaults to make `ExtendedSource` always monotonic and implement a `SymmetricSource` that is monotonic and symmetric.

Reply:
I wouldn't be opposed to that, but it may be overkill. `ExtendedSource` has the option of switching two specific constraints on or off. That is self-contained. What you suggested is to allow a user-specified constraint to be attached to a parameter. That constraint is potentially not contained in scarlet code and cannot be pickled.

Reply:
Fair enough. We can leave it as is.